### PR TITLE
FV: Guardian Can Always Initiate Recovery

### LINKED ISSUE #18: FV: Guardian can always initiate recovery
## Intention

> A guardian should always be able to initiate a recovery for the safe it is assigned.

### PR DESCRIPTION
**Pull request by remedcu**
Fixes #18

This PR adds a single rule which checks that under ideal conditions, the `confirmRecovery(...)` can always be called by a guardian to either initiate/execute the recovery process.

### CODE REVIEW

**Comment by mmv08** [comment]
Could you please attach a link to a successful run? It seems like the CI isn't run for this branch

**Comment by remedcu** [comment]
Here you go: https://prover.certora.com/output/950385/6aa7092ae38d400897835a26df4fa01a?anonymousKey=357b9b7adc4843d0003734a7d64005b56a3059ab

**Review by mmv08 — APPROVED** [review]

File: certora/specs/SocialRecoveryModule.spec [hunk 1]
```diff
@@ -158,3 +162,34 @@ rule revokeGuardianRevertIfAddressNotGuardian(env e, address otherAccount, addre
...
+rule confirmRecoveryCanAlwaysBeInitiatedByGuardian(env e, address guardian, address[] newOwners, uint256 newThreshold, bool execute) {
+
+    require newThreshold > 0;
+    require newThreshold <= newOwners.length;
+
+    require e.msg.value == 0;
+    require e.msg.sender == guardian;
+    require currentContract.isGuardian(safeContract, guardian);
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+    require e.block.timestamp + currentContract.recoveryPeriod < max_uint64;
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]: Can this be:
```suggestion
    require e.block.timestamp + currentContract.recoveryPeriod <= max_uint64;
```

File: certora/specs/SocialRecoveryModule.spec [hunk 2]
```diff
@@ -158,3 +162,34 @@ rule revokeGuardianRevertIfAddressNotGuardian(env e, address otherAccount, addre
...
+    require currentContract.getRecoveryApprovals(safeContract, newOwners, newThreshold) == currentContract.threshold(safeContract);
+
+    currentContract.confirmRecovery@withrevert(e, safeContract, newOwners, newThreshold, execute);
+    bool isReverted = lastReverted;
+
+    assert !isReverted &&
+        currentContract.confirmedHashes[recoveryHash][e.msg.sender];
+    assert execute =>
+        to_mathint(currentContract.recoveryRequests[safeContract].executeAfter) == e.block.timestamp + currentContract.recoveryPeriod &&
+        currentContract.recoveryRequests[safeContract].newThreshold == newThreshold;
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]: Should we assert something about new owners?

File: certora/specs/SocialRecoveryModule.spec [hunk 3]
```diff
@@ -158,3 +162,34 @@ rule revokeGuardianRevertIfAddressNotGuardian(env e, address otherAccount, addre
...
+    require e.msg.sender == guardian;
+    require currentContract.isGuardian(safeContract, guardian);
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+    require e.block.timestamp + currentContract.recoveryPeriod < max_uint64;
+
+    uint256 nonce = currentContract.nonce(safeContract);
+    require nonce < max_uint256;
+
+    bytes32 recoveryHash = currentContract.encodeRecoveryDataHash(safeContract, newOwners, newThreshold, nonce);
+    require currentContract.recoveryRequests[safeContract].executeAfter == 0; // This ensures that the recovery is not already initiated.
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]: Same comment as in the other PR - some of these `require`s are the conditions for the "always" that we should document.

**Review by nlordell — COMMENTED** [review]
LGTM - just some small requests (add assertion about newOwners somewhere, if possible and documentation on the conditions for the "always")

File: certora/specs/SocialRecoveryModule.spec (hunk 3 above)
Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec]: Done 👍🏾

File: certora/specs/SocialRecoveryModule.spec (hunk 1 above)
Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec]: Nice catch, done 👍🏾

File: certora/specs/SocialRecoveryModule.spec (hunk 2 above)
Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec]: I added the length parameter check and a random index check in the owner array.

**Comment by remedcu** [comment]
Latest run: https://prover.certora.com/output/950385/abe8e2a7cdb740a2b44668b2232e625f?anonymousKey=5a15718b0dc19d0889fa35b20c9ce82c88daa24f

File: certora/specs/SocialRecoveryModule.spec [hunk 4]
```diff
@@ -214,21 +214,25 @@ rule revokeGuardianRevertPossibilities(env e, address otherAccount, address prev
...
+    // No ether should be sent as part of this function call, and the caller should be a guardian.
     require e.msg.value == 0;
     require e.msg.sender == guardian;
     require currentContract.isGuardian(safeContract, guardian);
-    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
-    require e.block.timestamp + currentContract.recoveryPeriod < max_uint64;
 
+    requireGuardiansLinkedListIntegrity();
+
+    // Nonce and timestamp + recovery period should not overflow.
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]: Maybe qualify that this isn't realistically expected to happen under normal use.

File: certora/specs/SocialRecoveryModule.spec [hunk 5]
```diff
@@ -207,3 +211,42 @@ rule revokeGuardianRevertPossibilities(env e, address otherAccount, address prev
         !safeContract.isModuleEnabled(currentContract) ||
         guardianStorageContract.entries[safeContract].guardians[prevGuardian] != otherAccount;
 }
+
+// This rule verifies that the guardian can always initiate recovery considering some ideal conditions.
+rule confirmRecoveryCanAlwaysBeInitiatedByGuardian(env e, address guardian, address[] newOwners, uint256 newThreshold, bool execute, uint256 index) {
+    require index < newOwners.length;
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec:222]:
```suggestion
    // Index must be valid.
    require index < newOwners.length;
```

File: certora/specs/SocialRecoveryModule.spec [hunk 6]
```diff
@@ -207,3 +211,42 @@ rule revokeGuardianRevertPossibilities(env e, address otherAccount, address prev
...
+    currentContract.confirmRecovery@withrevert(e, safeContract, newOwners, newThreshold, execute);
+    bool isReverted = lastReverted;
+
+    assert !isReverted &&
+        currentContract.confirmedHashes[recoveryHash][e.msg.sender];
+    assert execute =>
+        to_mathint(currentContract.recoveryRequests[safeContract].executeAfter) == e.block.timestamp + currentContract.recoveryPeriod &&
+        currentContract.recoveryRequests[safeContract].newThreshold == newThreshold &&
+        currentContract.recoveryRequests[safeContract].newOwners.length == newOwners.length &&
+        currentContract.recoveryRequests[safeContract].newOwners[index] == newOwners[index];
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec:256]: I think this trick can be used in the `requireGuardiansLinkedListIntegrity` to require that `isGuardian(address) => getGuardians()[index] == address`.

File: certora/specs/SocialRecoveryModule.spec (hunk 5 above)
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]: ubernit: Index isn't part of the rule parameterization, but rather used internally for verifying integrity. Just a small suggestion so it doesn't "leak" to the rule signature.

```suggestion
rule confirmRecoveryCanAlwaysBeInitiatedByGuardian(env e, address guardian, address[] newOwners, uint256 newThreshold, bool execute) {
    uint256 index;
    require index < newOwners.length;
```

**Review by nlordell — APPROVED** [review]

File: certora/specs/SocialRecoveryModule.spec (hunk 5 above)
Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec]: Done 👍🏾

Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec:222]: Done 👍🏾

File: certora/specs/SocialRecoveryModule.spec (hunk 4 above)
Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec]: Done 👍🏾

File: certora/specs/SocialRecoveryModule.spec (hunk 6 above)
Comment (remedcu) [reply certora/specs/SocialRecoveryModule.spec:256]: Added in PR #21

**Comment by remedcu** [comment]
Latest run: https://prover.certora.com/output/950385/0012c310d73744b9a9171d633415b9fe?anonymousKey=90d33270db54917a731dbb9107b391b5081a86b5

**Pull request merged**